kube-runtime/src/scheduler.rs

```diff
@@ -126,6 +126,53 @@ impl<'a, T: Hash + Eq + Clone, R> SchedulerProj<'a, T, R> {
             }
         }
     }
+
+    /// Attempt to retrieve a message from queue and mark it as pending.
+    pub fn pop_queue_message_into_pending(&mut self, cx: &mut Context<'_>) -> Poll<T> {
+        loop {
+            match self.queue.poll_expired(cx) {
+                Poll::Ready(Some(msg)) => {
+                    let msg = msg.into_inner();
+                    self.pending.insert(msg);
+                }
+                Poll::Ready(None) | Poll::Pending => break Poll::Pending,
+            }
+        }
+    }
+}
+
+/// See [`Scheduler::hold`]
+pub struct Hold<'a, T, R> {
+    scheduler: Pin<&'a mut Scheduler<T, R>>,
+}
+
+impl<'a, T, R> Stream for Hold<'a, T, R>
+where
+    T: Eq + Hash + Clone,
+    R: Stream<Item = ScheduleRequest<T>>,
+{
+    type Item = T;
+
+    #[allow(clippy::match_wildcard_for_single_variants)]
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+        let mut scheduler = this.scheduler.as_mut().project();
+
+        loop {
+            match scheduler.requests.as_mut().poll_next(cx) {
+                Poll::Ready(Some(request)) => scheduler.schedule_message(request),
+                Poll::Ready(None) => return Poll::Ready(None),
+                Poll::Pending => break,
```

Comment:
when does the `Poll::Pending` match arm happen here? it also does not seem to be tested

Reply:
When the scheduler has no incoming requests.